src/lightly_train/_commands/common_helpers.py

```diff
@@ -391,18 +392,82 @@ def get_dataset_temp_mmap_path(
     """
     out_hash = get_sha256(f"{data}-{distributed_helpers.get_node_rank() or 0}")
     mmap_filepath = (cache.get_data_cache_dir() / out_hash).with_suffix(".mmap")
+    ref_count_filepath = mmap_filepath.with_suffix(".ref_count")
+
     mmap_filepath.parent.mkdir(parents=True, exist_ok=True)
 
-    reuse_file = Env.LIGHTLY_TRAIN_MMAP_REUSE_FILE.value
     try:
-        # Delete the file if it already exists from a previous run.
-        if not reuse_file and distributed_helpers.is_local_rank_zero():
-            _unlink_and_ignore(mmap_filepath)
+        # Increment reference count atomically
+        if distributed_helpers.is_local_rank_zero():
+            _increment_ref_count(ref_count_filepath)
 
         yield mmap_filepath
     finally:
-        if not reuse_file and distributed_helpers.is_local_rank_zero():
-            _unlink_and_ignore(mmap_filepath)
+        # Decrement reference count and cleanup if zero
+        if distributed_helpers.is_local_rank_zero():
+            _decrement_and_cleanup_if_zero(mmap_filepath, ref_count_filepath)
+
+
+def _increment_ref_count(ref_file: Path) -> None:
+    lock_file = ref_file.with_suffix(".lock")
+
+    # Retry logic for Python 3.8 robustness
+    max_retries = 10
+    for attempt in range(max_retries):
+        try:
+            with FileLock(lock_file):
+                # Ensure file exists within the lock to avoid race conditions
+                ref_file.touch()
+                with open(ref_file, "r+") as f:
+                    count = int(f.read() or "0")
+                    f.seek(0)
+                    f.write(str(count + 1))
+                    f.truncate()
+                break
+        except (OSError, ValueError):
+            if attempt == max_retries - 1:
+                raise
+            # Brief delay before retry, especially helpful on Python 3.8
+            time.sleep(0.001 * (attempt + 1))
+
+
+def _decrement_and_cleanup_if_zero(mmap_file: Path, ref_file: Path) -> None:
+    try:
+        should_cleanup = False
+        lock_file = ref_file.with_suffix(".lock")
+
+        # Retry logic for Python 3.8 robustness
+        max_retries = 10
+        for attempt in range(max_retries):
+            try:
+                with FileLock(lock_file):
+                    with open(ref_file, "r+") as f:
+                        count = int(f.read() or "1") - 1
+                        if count <= 0:
+                            should_cleanup = True
+                        else:
+                            f.seek(0)
+                            f.write(str(count))
+                            f.truncate()
```

Comment:
```suggestion
                with FileLock(lock_file):
                    with open(ref_file, "r+") as f:
                        count = int(f.read() or "1") - 1
                        f.seek(0)
                        f.write(str(count))
                        f.truncate()
                        if count <= 0:
                            should_cleanup = True
                    if not Env.LIGHTLY_TRAIN_MMAP_REUSE_FILE.value:
                        # Remove mmap file only if we are not reusing it
                        _unlink_and_ignore(mmap_file)  
```
You have to unlink and ignore the files within the same `with FileLock(...)` statement. Once you release the lock the `should_cleanup` condition might not be true anymore as another process might already change something. I would also not remove the ref count file as it uses zero space on the disk.